Restore Seperate test builds, but fix broken pipe

sdmaclea opened the pull request.

sdmaclea committed:
Build coreclr targetGeneric tests separately for CI (#34790)

* Build CoreClr tests separately

* Build coreclr targetGeneric tests separately for CI
* Build coreclr targetGeneric tests separately in runtime.yml
* Distinguish generic build based on Libraries config
* Build OSX release libraries when CoreCLR changed
* Always use tar.gz compression for generic tests

* Mark all CoreCLR Interop/COM projects OsSpecific

sdmaclea committed:
Fix duplicate job definition

safern commented on eng/pipelines/common/templates/runtimes/run-test-job.yml:
```diff
@@ -57,6 +58,11 @@ jobs:
         - '${{ parameters.runtimeFlavor }}_common_test_build_p0_${{ parameters.managedTestBuildOsGroup }}${{ parameters.managedTestBuildOsSubgroup }}_${{ parameters.archType }}_${{parameters.buildConfig }}'
       - ${{ if ne(parameters.testGroup, 'innerloop') }}:
         - '${{ parameters.runtimeFlavor }}_common_test_build_p1_${{ parameters.managedTestBuildOsGroup }}${{ parameters.managedTestBuildOsSubgroup }}_${{ parameters.archType }}_${{parameters.buildConfig }}'
+      - ${{ if eq(parameters.testBuildPhased, true) }}:
```
Why do we have to still depend on the non-phased build if this is to true?

sdmaclea replied:
`testBuildPhased` here means that the tests were built in two phases.  `TargetGeneric` and `TargetSpecific` tests. We join them here by unzipping into the same directory.

safern replied:
So if testBuildPhased==true we still depend on: 
`- '${{ parameters.runtimeFlavor }}_common_test_build_p0_${{ parameters.managedTestBuildOsGroup }}${{ parameters.managedTestBuildOsSubgroup }}_${{ parameters.archType }}_${{parameters.buildConfig }}'`

sdmaclea replied:
Yes

sdmaclea replied:
@trylek And I discussed this briefly during the original review.  

We may eventually refactor and build the `TargetSpecific` tests as part of `run-test-job`.  He was suggesting collecting stats to determine whether the extra build legs added value or not.

safern replied:
That sounds like a reasonable, interesting idea.

trylek replied:
In fact, I was mostly worried (based on earlier discussions with @jashook) that adding too many legs can be counterproductive by increasing the load on machine allocation and provisioning. I however think that having the specific test build legs initially separate is probably a good idea to get an initial reading on the perf characteristics.

safern replied:
Right, but if we refactor and just have build-test legs for the generic tests and then the target specific builds are done at the run-test-job, before sending to helix, we would be reducing the number of legs we will have by the number of target specific tests we currently have, so I think it would improve perf and reliability.

sdmaclea replied:
I suspect you are right.

+ We have reduce the target specific build by at least 90%.
+ The extra leg appears expensive (maybe 50% overhead). I was seeing target specific 10 minute builds in the pipelines.  I don't think it was that long on my local machine 
+ We should only have to rerun a given leg < 1% of the time (if we achieve 95% success).
+ Fewer legs should reduce infra burden, so we could se benefits elsewhere.
+ Seems like the savings could easily amortize the cost.

Filtering to build the TargetSpecific tests is currently slow.  But if we make msbuild improvements this may go away.

safern commented on eng/pipelines/coreclr/ci.yml:
```diff
@@ -92,6 +92,16 @@ jobs:
+    testGroup: outerloop
+    jobParameters:
+      liveLibrariesBuildConfig: Release
+      testBuildPhase: targetGeneric
```
nit, add an extra line in between two -template sections.

safern commented on eng/pipelines/runtime.yml:
```diff
@@ -368,6 +368,13 @@ jobs:
     - Windows_NT_arm64
     - Windows_NT_x86
+    # Generic test build always needs Release libraries.
+    - ${{ and(
```
This condition is not correct. You just need to add this condition:

```yml
- ${{ if eq(variables['isFullMatrix'], false) }}:
  - OSX_x64
```

That will cause to only run this on PRs.

safern commented on eng/pipelines/runtime.yml:
```diff
@@ -385,6 +385,24 @@ jobs:
+    jobTemplate: /eng/pipelines/libraries/build-job.yml
+    buildConfig: Release
+    platforms:
+    - ${{ if ne(variables['debugOnPrReleaseOnRolling'], 'Release') }}:
```
Instead use, `variables['isFullMatrix']` -> true, means Rolling build, false means PR.

sdmaclea replied:
Interesting...  I used this incantation because of line 377 above.  I can switch, but this was obvious to me.

sdmaclea commented:
I just forced pushed...

safern commented on eng/pipelines/runtime.yml:
```diff
@@ -576,6 +594,20 @@ jobs:
+      condition: >-
+        or(
+          eq(dependencies.checkout.outputs['SetPathVars_coreclr.containsChange'], true),
+          eq(variables['isFullMatrix'], true))
```
Please add an empty line in between template declarations.

safern commented on eng/pipelines/runtime.yml:
```diff
@@ -597,6 +630,21 @@ jobs:
+        and(ne(variables['debugOnPrReleaseOnRolling'], 'Release'),
+          or(
+            eq(dependencies.checkout.outputs['SetPathVars_coreclr.containsChange'], true),
+            eq(variables['isFullMatrix'], true)))
```
ditto.

safern commented on eng/pipelines/runtime.yml:
```diff
@@ -576,6 +594,20 @@ jobs:
 # CoreCLR Test builds using live libraries release build
 # Only when CoreCLR is changed
 #
+- template: /eng/pipelines/common/platform-matrix.yml
```
Could you explain why we need to add this pipeline? This seems exactly equal as what you're adding in L633.

sdmaclea replied:
This is builds the `targetGeneric` tests which were disabled in the template at line L611 by L623.

sdmaclea replied:
L605 is different than L642. So Not adding this would have resulted in changing what we ran.

safern replied:
I guess my question is, why do we need to build the tests against a debug and against a release build of libraries? I need more context on this change, could you elaborate what the requirements are here?

sdmaclea replied:
My task was to split the building of tests during CI into `targetSpecific` tests and `targetGeneric` tests.  Build the `targetGeneric` tests on OSX.  Continue to run all tests on the platforms where they were currently running.

I think there was an issue here too.  I fixed it in my latest commit.

sdmaclea replied:
The rationale was we didn't need to build all the tests on all the platforms.  We only needed to build platform specific tests on all platforms.  OSX was chosen because it has historically finished building tests first.

safern replied:
Ok, I get that, and why do the generic tests need a release libraries build? That is the part that I don't get.

safern replied:
Right now you're adding three jobs. 
1. Libraries build Release
2. CoreCLR build tests (targetGeneric) using a Release libraries build (both on PR and CI).
3. CoreCLR build tests (targetGeneric) using a Debug libraries build (only on PR).

What I don't get why do we need these? As I understand this, we just need to add 1 build here:
1. CoreCLR build tests (targetGeneric) uses Debug on PR and Release on CI libraries.

sdmaclea replied:
That would change the test coverage.  We do not currently run the set of tests you describe.

The existing YAML looks like this.

```YAML
#
# CoreCLR Test builds using live libraries release build
# Only when CoreCLR is changed
#
- template: /eng/pipelines/common/platform-matrix.yml
  parameters:
    jobTemplate: /eng/pipelines/common/templates/runtimes/build-test-job.yml
    buildConfig: checked
    platforms:
    - Linux_arm 
    - Windows_NT_x86
    - Windows_NT_arm
    - Windows_NT_arm64
    jobParameters:
      testGroup: innerloop
      liveLibrariesBuildConfig: Release
      condition: >-
        or(
          eq(dependencies.checkout.outputs['SetPathVars_coreclr.containsChange'], true),
          eq(variables['isFullMatrix'], true))

#
# CoreCLR Test builds using live libraries debug build
# Only when CoreCLR is changed
#
- template: /eng/pipelines/common/platform-matrix.yml
  parameters:
    jobTemplate: /eng/pipelines/common/templates/runtimes/build-test-job.yml
    buildConfig: checked
    platforms:
    - OSX_x64
    - Linux_x64
    - Linux_arm64
    - Windows_NT_x64
    jobParameters:
      testGroup: innerloop
      liveLibrariesBuildConfig: ${{ variables.debugOnPrReleaseOnRolling }}
      condition: >-
        or(
          eq(dependencies.checkout.outputs['SetPathVars_coreclr.containsChange'], true),
          eq(variables['isFullMatrix'], true))
```

As I read it we always run `Release` version of libraries on a subset of our platforms.

Because I need to build `Debug` and `Release` in some configurations, I need the three jobs.

safern replied:
but what does libraries build config matter when you're building the tests? You can build against whatever libraries config, the run-test-job will download libraries again and use that to run, no? 

https://github.com/dotnet/runtime/blob/master/eng/pipelines/common/templates/runtimes/run-test-job.yml#L181

sdmaclea replied:
I believe you are right.  I hadn't looked closely enough at the run job to realize it didn't matter.

I'll remove the unnecessary legs.

Thanks for explaining the root cause why it wasn't necessary.

It is amazing to me how difficult communication can be 😉

safern replied:
No problem. Yeah and specially with so many configurations an templates it’s hard to keep track what is needed and what’s not.

Thanks for elaborating on the reason why you were doing this. Let’s remove the unnecessary extra leg and see if everything goes green which I think it should as run-test-job should download the right libraries build. :fingerscrossed:

sdmaclea replied:
I pushed the change.

sdmaclea committed:
Whitespace feedback

sdmaclea committed:
Condition feedback

Use isFullMatrix rather than debugOnPrReleaseOnRolling
Simplify logic
Fix another duplicate job case

sdmaclea committed:
Assume debug/release libraries doesn't matter during test build

trylek commented on eng/pipelines/common/templates/runtimes/run-test-job.yml:
```diff
@@ -189,6 +195,13 @@ jobs:
+            unpackFolder: '$(managedTestArtifactRootFolderPath)'
+            artifactFileName: '$(managedGenericTestArtifactName).tar.gz'
+            artifactName: '$(managedGenericTestArtifactName)'
+            displayName: 'generic managed test artifacts'
```
This might be worth 1-2 more comments explaining the generic vs. specific test management. It would be probably too nitpicky to alternate the displayName of the pre-existing download between "managed test artifacts" and "generic managed test artifacts" but we should try to keep these scripts easy to understand and possibly refactor.

trylek reviewed (approved):
Looks even better to me than the last time, thank you!

sdmaclea committed:
Add comments
